### PR TITLE
global -p option

### PR DESCRIPTION
**Pull request by drwebb**
@snoyberg Any suggestions for naming these options: `--local-bin-path` seems to make sense, but I'm not a fan of `-p`. Also note that the invocation will be a lot more natural once #519 is implemented.

### CODE REVIEW

**Comment by snoyberg** [comment]
I'm -1 on the `-p` shorthand. I'd say `--install-path` would be better, but perhaps @manny-fp would have input. Frankly, even `--path` seemed OK, but now that it's a global option I can see making it a bit more explicit would be worthwhile.

File: src/Stack/Config.hs [hunk 1]
```diff
@@ -146,9 +146,17 @@ configFromConfigMonoid configStackRoot mproject configMonoid@ConfigMonoid{..} =
...
 
-     configLocalBin <- do
-        localDir <- liftIO (getAppUserDataDirectory "local") >>= parseAbsDir
-        return $ localDir </> $(mkRelDir "bin")
+     configLocalBin <- case configMonoidLocalBin of
+       Nothing -> do
+           localDir <- liftIO (getAppUserDataDirectory "local") >>= parseAbsDir
+           return $ localDir </> $(mkRelDir "bin")
+       Just userPath -> do
+           tryPath <- try (liftIO $ canonicalizePath userPath >>= parseAbsDir)
```
Comment (snoyberg) [inline src/Stack/Config.hs]: This won't work if the directory doesn't already exist. Possible solution: call `createDirectoryIfMissing True userPath` first. Not sure if it's worth it.

File: src/Stack/Config.hs [hunk 2]
```diff
@@ -146,9 +146,17 @@ configFromConfigMonoid configStackRoot mproject configMonoid@ConfigMonoid{..} =
...
+           case tryPath of
+               Left (_ :: SomeException) ->
```
Comment (snoyberg) [inline src/Stack/Config.hs]: This shouldn't be a `SomeException`, but an `IOException`. Also, displaying that information to the user would be a good thing.

**Comment by drwebb** [comment]
Summary of commit be8dbb1:

Dropped `-p` short option. I like `--path` as well but `--local-bin-path` is a little more explicit that we are just copying the binaries to that location. I'm still up to changing to `--install-path` though might imply other things than just the bin. As it it `--local-bin-path` is clear at the expense of looking a little unconventional.  

I am opposed to employing `createDirectoryIfMissing`. We want the user to see something like
```
λ church stack → λ git userlocalbin* → ~/.local/bin/stack --local-bin-path ./binz install
Could not locate user specified directory "./binz"
IOException: ./binz: canonicalizePath: does not exist (No such file or directory)
```

Basically fail fast and hard on a typo, not silently doing what the user doesn't want.

The other change was to match only on IOException, which is the only thing that I think canonicalizePath can throw. Looking at the code for Path, it can throw a PathParseException which I will go ahead and cover.

**Comment by drwebb** [comment]
Summary of 55a5476:
Also catching PathParseExceptions right there, never been able to trigger it (canonicalizePath seems to catch any nastyness) though.

**Comment by drwebb** [comment]
@snoyberg: I talked with @manny-fp about the naming and he also liked `--local-bin-path` as one can already use `stack path --local-bin-path`. I should also make sure the new field in the `stack.yaml` will follow the same convention, so if this looks good to you I'll handle the merge.

**Comment by snoyberg** [comment]
@manny-fp Please merge when you're ready, I'm fine with this PR.

File: src/Stack/Config.hs [hunk 3]
```diff
@@ -146,9 +147,23 @@ configFromConfigMonoid configStackRoot mproject configMonoid@ConfigMonoid{..} =
...
-        return $ localDir </> $(mkRelDir "bin")
+     configLocalBin <-
+         case configMonoidLocalBin of
+             Nothing -> do
+                 localDir <- liftIO (getAppUserDataDirectory "local") >>= parseAbsDir
+                 return $ localDir </> $(mkRelDir "bin")
+             Just userPath ->
+                 (liftIO $ canonicalizePath userPath >>= parseAbsDir)
+                 `catches`
+                 [Handler (\(ex :: IOException) -> error
```
Comment (manny-fp) [inline src/Stack/Config.hs]: Should not be using `error`, instead throw a new [ConfigException](https://github.com/commercialhaskell/stack/blob/fc963d1f0faded90fb8ca7e5375613a0b11b67c4/src/Stack/Types/Config.hs#L515) constructor.

File: src/Stack/Types/Config.hs [hunk 4]
```diff
@@ -500,6 +504,7 @@ instance FromJSON ConfigMonoid where
          configMonoidExtraIncludeDirs <- obj .:? "extra-include-dirs" .!= Set.empty
          configMonoidExtraLibDirs <- obj .:? "extra-lib-dirs" .!= Set.empty
          configMonoidConcurrentTests <- obj .:? "concurrent-tests"
+         configMonoidLocalBin <- obj .:? "local-bin"
```
Comment (manny-fp) [inline src/Stack/Types/Config.hs]: Should this be `"local-bin-path"` for consistency with the command-line argument name?
